File: Problem Set 1/code/src/p01b_logreg.py
import numpy as np
import util

from linear_model import LinearModel


def main(train_path, eval_path, pred_path):
    """Problem 1(b): Logistic regression with Newton's Method.

    Args:
        train_path: Path to CSV file containing dataset for training.
        eval_path: Path to CSV file containing dataset for evaluation.
        pred_path: Path to save predictions.
    """
    x_train, y_train = util.load_dataset(train_path, add_intercept=True)
    x_eval, y_eval = util.load_dataset(eval_path, add_intercept=True)
    # *** START CODE HERE ***
    ## train the model
    logreg = LogisticRegression()
    logreg.fit(x_train, y_train)

    y_pred = logreg.predict(x_eval)
    # plot decision boundary
    util.plot(x=x_train, y=y_train, theta=logreg.theta, save_path='output/p01b{0}.png'.format(pred_path[16:-4]),
              x_eval=x_eval, y_pred=y_pred)
    np.savetxt(pred_path, y_pred, fmt='%d')
    # *** END CODE HERE ***

class LogisticRegression(LinearModel):
    """Logistic regression with Newton's Method as the solver."""

    def fit(self, x, y):
        """Run Newton's Method to minimize J(theta) for logistic regression.

        Args:
            x: Training example inputs. Shape (m, n).
            y: Training example labels. Shape (m,).
        """
        # *** START CODE HERE ***
        
        ## init theta with Shape of (n, )
        m, n = x.shape
        self.theta = np.zeros(n)

        ## apply newton's method
        n_iter = 0
        while n_iter < self.max_iter:

            h_x = util.sigmoid(x.dot(self.theta))
            gradient_J = -x.T.dot(y - h_x) / m
            H_J = (x.T * h_x * (1- h_x)).dot(x) / m

            ## update theta & check if converge
            step = np.linalg.inv(H_J).dot(gradient_J)
            self.theta -= step
            if np.linalg.norm(step, ord=1) < self.eps: break
            n_iter += 1
        
        # *** END CODE HERE ***

    def predict(self, x, p=False):
        """Make a prediction given new inputs x.

        Args:
            x: Inputs of shape (m, n).
            p: if return prob of the output

        Returns:
            Outputs of shape (m,).
        """
        # *** START CODE HERE ***
        h_x_opt = util.sigmoid(x.dot(self.theta))
        if p:
            return h_x_opt
        else:
            return np.where(h_x_opt >= 0.5, 1, 0)
        # *** END CODE HERE ***

File: Problem Set 1/Code/src_cpp/src_cpp/p01b_logreg.cpp
/*
source file to apply logistic regression
*/
#include <armadillo>
#include "p01b_logreg.hpp"
#include "util.hpp"

using namespace std;
using namespace arma;

// constructors
LogisticRegression::LogisticRegression() : LinearModel(){};
LogisticRegression::LogisticRegression(const LogisticRegression& source) : LinearModel(source) {};
LogisticRegression::LogisticRegression(mat& theta, const double& step_size, const int& max_iter, const double& eps) : LinearModel(theta, step_size, max_iter, eps) {};
LogisticRegression::~LogisticRegression() {};

// assignment operator
LogisticRegression& LogisticRegression::operator=(const LogisticRegression& source) {
	if (this == &source) {
		cout << "self-assignment checked";
	}
	else {
		LinearModel::operator=(source);
	}
	return *this;
}

// functions
const void LogisticRegression::fit(const mat& x, const mat& y) {
	/*
	Run Newton's Method to minimize J(theta) for logistic regression.
	Args:
	x: Training example inputs.Shape(m, n).
	y : Training example labels.Shape(m, ).
	*/
	arma::uword m = x.n_rows, n = x.n_cols;
	theta = arma::zeros(n); // initate theta with zeros, shape of (m, 1)
	
	// apply newton's method
	int n_iter = 0;
	
	while (n_iter < max_iter) {
		mat h_x = util::sigmoid(x * theta);

		mat gradient_J = -x.t() * (y - h_x) / m;
		mat H_J = (x.each_col() % (h_x % (1 - h_x))).t() * x / m;
		// update theta & check if coverge
		vec step = arma::inv(H_J) * gradient_J;
		theta -= step;
		if (arma::norm(step, 1) < eps) { break; };
		n_iter += 1;
	}
	return;
}

const mat LogisticRegression::predict(const mat& x, const bool& p) {
	/*
	Make a prediction given new inputs x.
	Args:
	x: Inputs of shape(m, n).
	p : if return prob of the output

	Returns :
	Outputs of shape(m, 1);
	*/

	// hypothestic function results with trained theta
	mat h_x_opt = util::sigmoid(x * theta);
	arma::uword m = h_x_opt.n_rows;

	if (p) { return h_x_opt; }
	else {
		mat ks(m, 1);
		for (int i = 0; i < m; i++) {
			if (h_x_opt[i] >= 0.5) { ks[i] = 1; }
			else { ks[i] = 0; };
		}
		return ks;
	}
}

// executive function of p01b_logreg
void p01b_logreg(string dataset) {
	// loadata set
	std::string root, path_train, path_eval, savedr;
	root = "C:\\Users\\YB\\Documents\\GitHub\\Standford_CS229\\Problem Set 1\\Code\\data\\";
	savedr = "C:\\Users\\YB\\Documents\\GitHub\\Standford_CS229\\Problem Set 1\\Code\\src_cpp\\src_cpp\\output\\";
	path_train = dataset + "_train.csv";
	path_eval = dataset + "_valid.csv";

	tuple<mat, mat> data_train = util::load_dataset(root + path_train);
	tuple<mat, mat> data_eval = util::load_dataset(root + path_eval);

	mat X_train, y_train, X_eval, y_eval;

	// Get train/eval data
	X_train = get<0>(data_train);
	y_train = get<1>(data_train);

	X_eval = get<0>(data_eval);
	y_eval = get<1>(data_eval);

	// train dataset1, predict, save
	LogisticRegression logreg;
	logreg.fit(X_train, y_train);

	mat y_pred;
	y_pred = logreg.predict(X_eval);
	y_pred.save(savedr + "p01b_pred_logreg_" + dataset + ".csv", arma::arma_ascii);

	return;
}